fairscale/nn/data_parallel/fully_sharded_data_parallel.py

```diff
@@ -383,17 +384,18 @@ def state_dict(self, *args, **kwargs):  # type: ignore
         .. warning:: This needs to be called on all ranks, since synchronization
             primitives will be used.
         """
-        torch.cuda.synchronize()
-        self._lazy_init()
-        self._rebuild_full_params()
-        self._all_buffers_to(dtype=torch.float32)  # Buffers dtype stays consistent with parameters.
-        state_dict = self.module.state_dict(*args, **kwargs)
-        # We don't free the params after generating the state dict, since
-        # freeing is done in-place (via the Storage) and would corrupt the
-        # returned state dict. However, we need to maintain the invariant that
-        # p.data corresponds to the FP32 param shard, so we do that here.
-        self._use_fp32_param_shard()
-        self._all_buffers_to(dtype=self.compute_dtype)
+        with self.summon_full_params():
+            # Buffers dtype stays consistent with parameters.
+            self._all_buffers_to(dtype=torch.float32)
+
+            state_dict = self.module.state_dict(*args, **kwargs)
+            # We copy the state_dict since full param will be free after
+            # we exit the summon_full_params() context.
+            # TODO (Min): if the state is huge, the copy can OOM.
+            state_dict = copy.deepcopy(state_dict)
```

Comment:
@myleott @sshleifer let me know if this deepcopy is OK. It seems that for very huge model this would be bad, but it restores invariant of freeing full params after use. For huge models, state checkpointing probably needs more fundamental work anyway.

Reply:
Ah, for very large models we will use `local_state_dict` which is not affected by this, so all good. There is some inefficiency here, but we can revisit if it becomes a problem.